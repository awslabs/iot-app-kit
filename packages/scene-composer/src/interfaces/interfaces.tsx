/**
 * This file contains public interfaces of the scene content the SceneComposer exposes.
 * It is different from the serialization format from the GLTF Toolkit package.
 * These interfaces are used to interact with the SceneComposer instance at
 * runtime, thus the data structure is slightly different from the GLTF Toolkit.
 */

import { ReactElement } from 'react';

import { ISceneNodeInternal } from '../store/Store';

import { ISceneComponent, KnownComponentType } from './components';
import { IRuleBasedMap, ITransform, ITransformConstraint, Vector3 } from './dataTypes';

/************************************************
 * Scene document, entity, component models
 ************************************************/

export interface ISceneDocumentSnapshot {
  /**
   * Serialize the current document snapshot into a document string with the
   * given format version.
   *
   * Note that this interface is not optimized for very large scene, which
   * we may introduce new APIs later to use stream API to handle the output.
   *
   * TODO: do we want to allow updating the document version?
   * [Option 1]: Externalize the document version instea of keeping the doucment version in the document itself.
   * [Option 2]: Allow passing in a parameter to set the document version.
   * [Option 3]: Automatically increment the document version by 1 based on the original input document version.
   *
   * We'll go with option 1 for now for simplicity.
   *
   * TODO: what should we return when the scene is empty?
   * [Option 1]: Return a falsy value, such as empty string.
   * [Option 2]: Return a valid scene with no entity.
   *
   * We'll go with option 2 for now.
   *
   * @param specVersion - version of the document format
   */
  serialize(specVersion: string, ...stringifyArgs): string;
}

export interface ISceneDocument {
  nodeMap: Record<string, ISceneNode>;
  ruleMap: Record<string, IRuleBasedMap>;
  rootNodeRefs: string[];
  unit?: string;
  version: string;
  properties?: Partial<Record<KnownSceneProperty, any>>;
  specVersion?: string;
}

export interface ISceneNode {
  // ref is optional when used in APIs to create nodes
  ref?: string;
  name?: string;
  transform?: ITransform;
  transformConstraint?: ITransformConstraint;
  components?: ISceneComponent[];
  parentRef?: string;
  childRefs?: string[];
  properties?: Partial<Record<'alwaysVisible' | 'matterportId' | 'hiddenWhileImmersive', boolean | string>>;
}

export enum KnownSceneProperty {
  BaseUrl = 'baseUrl',
  EnvironmentPreset = 'environmentPreset',
  DataBindingConfig = 'dataBindingConfig',
  ComponentSettings = 'componentSettings',
  MatterportModelId = 'matterportModelId',
  LayerIds = 'layerIds',
  SceneRootEntityId = 'sceneRootEntityId',
  TagCustomColors = 'tagCustomColors',
  FogSettings = 'fogSettings',
}

/************************************************
 * Scene Composer Control APIs
 ************************************************/

export type CameraControlMode = 'transition' | 'teleport';
export type CameraControlsType = 'orbit' | 'pan' | 'immersive' | 'pointerLock';
export type TransformControlMode = 'translate' | 'rotate' | 'scale';
export enum CameraViewAxisValues {
  Front = 'front',
  Back = 'back',
  Left = 'left',
  Right = 'right',
  Top = 'top',
  Bottom = 'bottom',
}

// an object ref or a precise camera setting
export type FixedCameraTarget = { position: Vector3; target: Vector3 };
export type NamedCameraTarget = string;
export type CameraTarget = NamedCameraTarget | FixedCameraTarget;

export type URIModifier = (uri: string) => string;

export type CreateErrorViewCallback = (error: Error) => ReactElement;

/************************************************
 * Scene Namespace
 ************************************************/

const IotTwinMakerNamespace = 'iottwinmaker.common';
export const IotTwinMakerNamespaceSeparator = ':';
export const IotTwinMakerIconNamespace = `${IotTwinMakerNamespace}.icon`;
export const IotTwinMakerColorNamespace = `${IotTwinMakerNamespace}.color`;
export const IotTwinMakerNumberNamespace = `${IotTwinMakerNamespace}.number`;
export const IotTwinMakerOpacityNamespace = `${IotTwinMakerNamespace}.opacity`;
/************************************************
 * Scene Resource
 ************************************************/

export enum SceneResourceType {
  Icon = 'Icon',
  Color = 'Color',
  Number = 'Number',
  Opacity = 'Opacity',
}

export interface SceneResourceInfo {
  type: SceneResourceType;
  value: string;
}

export interface AddingWidgetInfo {
  type?: KnownComponentType;
  node: ISceneNodeInternal;
}

export interface MeshStyle {
  color?: string | number;
  opacity?: number;
  transparent?: boolean;
}

export interface StyleTarget {
  dataBindingContext: unknown;
  style: MeshStyle;
}

export interface IFogSettings {
  colorHex: number;
  near: number;
  far: number;
}
