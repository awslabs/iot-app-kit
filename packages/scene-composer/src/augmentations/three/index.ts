export * from './Anchor';
export * from './Viewpoint';
