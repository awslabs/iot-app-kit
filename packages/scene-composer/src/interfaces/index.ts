export * from './assets';
export * from './interfaces';
export * from './components';
export * from './dataTypes';
export * from './dataBinding';
export * from './metricRecorder';
export * from './feature';
export * from './sceneViewer';
export * from './componentSettings';
