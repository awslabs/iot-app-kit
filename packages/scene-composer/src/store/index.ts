export * from './StoreOperations';
export * from './Store';
