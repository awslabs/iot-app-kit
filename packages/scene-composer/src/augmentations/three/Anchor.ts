import * as THREE from 'three';

import DebugLogger from '../../logger/DebugLogger';
import { DefaultAnchorStatus, SelectedAnchor } from '../../interfaces';

import { IVisual } from './visuals';

/**
 * The generic anchor object.
 */
export class Anchor extends THREE.Object3D {
  private log = new DebugLogger('Anchor');

  protected visualMap = new Map<string, IVisual>();
  protected _visualState: DefaultAnchorStatus | string = DefaultAnchorStatus.Info;
  protected _isSelected = false;

  public add(visualWithStateName: IVisual): this {
    const stateName = visualWithStateName.name;
    if (!stateName) {
      this.log?.error('A name must be provided for the visual. This is used to determine its appearance.');
      return this;
    }

    if (!visualWithStateName.visual) {
      this.log?.error('A visual must be provided.');
      return this;
    }

    // NOTE: Add the states directly, because event bubbling to the anchor level does not seem to occur on click
    //  and is instead caught by the WidgetVisual
    super.add(visualWithStateName.visual);

    this.visualMap.set(stateName, visualWithStateName);

    // Update the anchor visual with the current visual state
    this.updateVisualState();

    return this;
  }

  public set visualState(visualState: DefaultAnchorStatus | string) {
    this._visualState = visualState;
    this.updateVisualState();
  }

  public get visualState(): DefaultAnchorStatus | string {
    return this._visualState;
  }

  /**
   * Gets the anchor's selected state
   */
  public get isSelected(): boolean {
    return this._isSelected;
  }

  /**
   * Sets the anchor's selected state keying the visibility of the specially named Selected Visual
   * @param value true if the anchor is selected
   */
  public set isSelected(value: boolean) {
    this._isSelected = value;
    this.visualMap.get(SelectedAnchor)?.setVisible(value);
  }

  private updateVisualState() {
    this.visualMap.forEach((visual, name) => {
      visual.setVisible(name === this.visualState);
    });
  }
}
