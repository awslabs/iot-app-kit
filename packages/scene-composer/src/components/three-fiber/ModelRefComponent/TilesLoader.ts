import { useMemo } from 'react';
import { useThree } from '@react-three/fiber';

import { TwinMakerGLTFLoader } from '../../../three/GLTFLoader';
import { TilesRenderer } from '../../../three/tiles3d/TilesRenderer';
import { setupTwinMakerGLTFLoader } from '../../../three/loaderUtils';
import { createTwinMakerFetch } from '../../../utils/TwinMakerBrowserUtils';
import { getGlobalSettings } from '../../../common/GlobalSettings';
import { URIModifier } from '../../../interfaces/interfaces';

import { fixNasaUriBug, setupTilesRenderer } from './TilesLoaderUtils';

/**
 * Load 3D Tiles. There are 2 steps to load tileset:
 *   1. Load tileset JSON
 *   2. For each frame, when TilesRenderer::update function is executed, the additional tileset or tiles will be loaded.
 * useTiles will only load #1, so it should be quick. We may add loading progress in it later, but not super urgent when
 * we are loading a small JSON file (compare with 1G of glTF).
 */
export function useTiles<T extends string>(path: T, uriModifier?: URIModifier) {
  const gl = useThree((state) => state.gl);
  const camera = useThree((state) => state.camera);
  const scene = useThree((state) => state.scene);

  const tilesRenderer = useMemo(() => {
    const getSceneObjectFunction = getGlobalSettings().getSceneObjectFunction;
    const tilesRenderer = new TilesRenderer(
      path,
      getSceneObjectFunction ? createTwinMakerFetch(getSceneObjectFunction) : undefined,
    );
    tilesRenderer.preprocessURL = (uri: string | URL) => {
      const uriString = fixNasaUriBug(uri);
      return uriModifier?.(uriString) ?? uriString;
    };

    tilesRenderer.loadSiblings = false;

    const loader = new TwinMakerGLTFLoader(tilesRenderer.manager);
    setupTwinMakerGLTFLoader(loader);

    tilesRenderer.manager.addHandler(/\.gltf$/, loader);
    tilesRenderer.onLoadTileSet = (ts) => setupTilesRenderer(tilesRenderer);
    tilesRenderer.setCamera(camera);
    tilesRenderer.setResolutionFromRenderer(camera, gl);
    return tilesRenderer;
  }, [camera, gl, scene]);

  return tilesRenderer;
}
