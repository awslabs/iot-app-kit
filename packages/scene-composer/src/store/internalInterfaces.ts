import { IErrorDetails } from '../common/errors';
import {
  SelectionChangedEventCallback,
  IAnchorComponent,
  ICameraComponent,
  IColorOverlayComponent,
  ILightComponent,
  IModelRefComponent,
  IMotionIndicatorComponent,
  IRuleBasedMap,
  IRuleStatement,
  ISceneComponent,
  ISceneDocument,
  ISceneNode,
  ITransform,
  ITransformConstraint,
  IValueDataBinding,
  IValueDataBindingProvider,
  OperationMode,
  ShowAssetBrowserCallback,
  URIModifier,
  KnownComponentType,
  WidgetClickEventCallback,
  ISubModelRefComponent,
  IDataOverlayComponent,
} from '../interfaces';
import { MapControls as MapControlsImpl, OrbitControls as OrbitControlsImpl } from '../three/OrbitControls';

export type ISerializationErrorDetails = IErrorDetails;

export interface IDeserializationResult {
  document?: ISceneDocumentInternal;
  errors: ISerializationErrorDetails[];
}

/******************************************************************************
 * Editor Data Model
 ******************************************************************************/

export enum DisplayMessageCategory {
  Error = 'Error',
  Warning = 'Warning',
  Info = 'Info',
}

export type CameraControlImpl = MapControlsImpl | OrbitControlsImpl;
export type TweenValueObject = { x: number; y: number; z: number };

export type CursorStyle = 'move' | 'edit';

export interface IDisplayMessage {
  category: DisplayMessageCategory;
  messageText: string;
  params?: Record<string, string>;
}

export interface IEditorConfig {
  operationMode?: OperationMode;
  uriModifier?: URIModifier;
  valueDataBindingProvider?: IValueDataBindingProvider;
  showAssetBrowserCallback?: ShowAssetBrowserCallback;
  onWidgetClick?: WidgetClickEventCallback;
  onSelectionChanged?: SelectionChangedEventCallback;
}

/******************************************************************************
 * Document Model
 ******************************************************************************/

export type ITransformInternal = ITransform;
export type IRuleBasedMapInternal = IRuleBasedMap;
export type IRuleStatementInternal = IRuleStatement;

export interface ISceneNodeInternal extends ISceneNode {
  // below fields are not nullable internally
  ref: string;
  name: string;
  transform: ITransformInternal;
  transformConstraint: ITransformConstraint;
  components: ISceneComponentInternal[];
  childRefs: string[];
  properties: Record<string, any>;
}

export interface ISceneDocumentInternal extends ISceneDocument {
  unit: string; // unit is not nullable internally, default to 'meter'
  ruleMap: Record<string, IRuleBasedMapInternal>;
  nodeMap: Record<string, ISceneNodeInternal>;
  componentNodeMap: { [type in KnownComponentType | string]?: Record<string, string[]> }; // record of node.ref to component.ref[]
}

export interface ISceneComponentInternal extends ISceneComponent {
  // unique id for the component for internal reference
  ref: string;
}

export interface IDataBoundSceneComponentInternal extends ISceneComponentInternal {
  valueDataBinding?: IValueDataBinding;
}

/******************************************************************************
 * Components
 ******************************************************************************/

export type ILightComponentInternal = ISceneComponentInternal & ILightComponent;

export type ICameraComponentInternal = ISceneComponentInternal & ICameraComponent;

export type IModelRefComponentInternal = ISceneComponentInternal & IModelRefComponent;

export type ISubModelRefComponentInternal = ISceneComponentInternal & ISubModelRefComponent;

export type IAnchorComponentInternal = IDataBoundSceneComponentInternal & IAnchorComponent;

export type IColorOverlayComponentInternal = IDataBoundSceneComponentInternal & IColorOverlayComponent;

export type IMotionIndicatorComponentInternal = ISceneComponentInternal & IMotionIndicatorComponent;

export type IDataOverlayComponentInternal = ISceneComponentInternal & IDataOverlayComponent;

/******************************************************************************
 * Type magic...
 ******************************************************************************/

export const isISceneNodeInternal = (item: ISceneNodeInternal | undefined): item is ISceneNodeInternal => {
  return !!item;
};

export const isISceneComponentInternal = (
  item: ISceneComponentInternal | undefined,
): item is ISceneComponentInternal => {
  return !!item;
};
