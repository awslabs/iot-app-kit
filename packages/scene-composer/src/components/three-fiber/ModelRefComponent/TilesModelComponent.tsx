import React from 'react';
import { ThreeEvent, useFrame } from '@react-three/fiber';
import { Object3D } from 'three';

import { MAX_CLICK_DISTANCE } from '../../../common/constants';
import useLifecycleLogging from '../../../logger/react-logger/hooks/useLifecycleLogging';
import { IModelRefComponentInternal, ISceneNodeInternal, useEditorState, useStore } from '../../../store';
import { acceleratedRaycasting, getComponentGroupName } from '../../../utils/objectThreeUtils';
import {
  findComponentByType,
  createNodeWithPositionAndNormal,
  findNearestViableParentAncestorNodeRef,
} from '../../../utils/nodeUtils';
import { useSceneComposerId } from '../../../common/sceneComposerIdContext';
import { KnownComponentType } from '../../../interfaces';
import { getIntersectionTransform } from '../../../utils/raycastUtils';

import { useTiles } from './TilesLoader';

interface TilesModelProps {
  node: ISceneNodeInternal;
  component: IModelRefComponentInternal;
}

export const TilesModelComponent: React.FC<TilesModelProps> = ({ node, component }: TilesModelProps) => {
  const sceneComposerId = useSceneComposerId();
  useLifecycleLogging('TilesModelComponent');
  const { getSceneNodeByRef } = useStore(sceneComposerId)((state) => state);
  const appendSceneNode = useStore(sceneComposerId)((state) => state.appendSceneNode);
  const uriModifier = useStore(sceneComposerId)((state) => state.getEditorConfig().uriModifier);
  const { isEditing, addingWidget, setAddingWidget, cursorLookAt } = useEditorState(sceneComposerId);

  // TODO: tilesRenderer holds "group" and it'll load tiles and B3DM/I3DM files dynanimcally, so we don't need
  //       to clone the model like what we did in GLTFModelComponent. However, if we found this assumption is
  //       wrong in the future, let's optimize from here.
  const tilesRenderer = useTiles(component.uri, uriModifier);

  // Enable optimized raycasting
  tilesRenderer.onLoadModel = (scene: Object3D) => {
    scene.traverse((obj: Object3D) => {
      acceleratedRaycasting(obj);
    });
  };

  useFrame(() => {
    tilesRenderer.update();
  });

  const handleAddWidget = (e: ThreeEvent<MouseEvent>) => {
    if (addingWidget) {
      const hierarchicalParent = findNearestViableParentAncestorNodeRef(e.eventObject);
      const hierarchicalParentNode = getSceneNodeByRef(hierarchicalParent?.userData.nodeRef);
      let physicalParent = hierarchicalParent;
      if (findComponentByType(hierarchicalParentNode, KnownComponentType.SubModelRef)) {
        while (physicalParent) {
          if (physicalParent.userData.componentTypes?.includes(KnownComponentType.ModelRef)) break;
          physicalParent = physicalParent.parent as THREE.Object3D<Event>;
        }
      }
      const { position } = getIntersectionTransform(e.intersections[0]);
      const newWidgetNode = createNodeWithPositionAndNormal(
        addingWidget,
        position,
        cursorLookAt,
        physicalParent,
        hierarchicalParent?.userData.nodeRef,
      );

      setAddingWidget(undefined);
      appendSceneNode(newWidgetNode);
      e.stopPropagation();
    }
  };

  const onClick = (e: ThreeEvent<MouseEvent>) => {
    if (e.delta <= MAX_CLICK_DISTANCE && isEditing() && addingWidget) {
      handleAddWidget(e);
    }
  };

  return (
    <group name={getComponentGroupName(node.ref, 'TILES_MODEL')} dispose={null}>
      <primitive object={tilesRenderer.group} onClick={onClick} />
    </group>
  );
};
