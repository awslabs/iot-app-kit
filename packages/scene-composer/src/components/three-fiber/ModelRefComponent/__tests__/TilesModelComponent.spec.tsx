import * as THREE from 'three';
import React from 'react';
import { render } from '@testing-library/react';
import { useFrame as mockUseFrame } from '@react-three/fiber';

import { TilesModelComponent } from '../TilesModelComponent';
import { KnownComponentType } from '../../../../interfaces';
import { IModelRefComponentInternal } from '../../../../store/internalInterfaces';
import { useTiles as mockUseTiles } from '../TilesLoader';

jest.mock('../TilesLoader', () => {
  return {
    useTiles: jest.fn(),
  };
});

jest.mock('@react-three/fiber', () => ({
  useFrame: jest.fn(),
}));

describe('TilesModelComponent', () => {
  const baseNode: any = {
    ref: 'mock-node',
  };
  const baseComponent: IModelRefComponentInternal = {
    ref: 'mock-comp',
    type: KnownComponentType.ModelRef,
    uri: 'mock/uri',
    modelType: 'Tiles3D',
  };
  const mockObject = new THREE.Mesh(new THREE.BoxGeometry(1, 2, 3), new THREE.MeshBasicMaterial({ color: 0xffff00 }));
  mockObject.name = 'mockObject';
  const baseScene = new THREE.Group();
  baseScene.add(mockObject);

  const setup = () => {
    jest.resetAllMocks();
  };

  beforeEach(() => {
    setup();
  });

  it('should render TilesModelComponent', () => {
    const mockUpdate = jest.fn();
    (mockUseTiles as jest.Mock).mockReturnValue({ update: mockUpdate, group: baseScene });

    const { container } = render(<TilesModelComponent node={baseNode} component={baseComponent} />);

    expect(container).toMatchSnapshot();
  });

  it('should update renderer every frame', () => {
    const mockUpdate = jest.fn();

    (mockUseTiles as jest.Mock).mockReturnValue({ update: mockUpdate, group: baseScene });
    (mockUseFrame as jest.Mock).mockImplementation((cb) => cb()); // basically, let's cheat and just run it as soon as useFrame is setup so we can verify it's internals

    render(<TilesModelComponent node={baseNode} component={baseComponent} />);

    expect(mockUpdate).toBeCalled();
  });
});
