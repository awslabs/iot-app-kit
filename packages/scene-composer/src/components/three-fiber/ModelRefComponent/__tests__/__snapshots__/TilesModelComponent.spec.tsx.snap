// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TilesModelComponent should render TilesModelComponent 1`] = `
<div>
  <group
    name="TILES_MODEL_COMPONENT_mock-node"
  >
    <primitive
      object="[object Object]"
    />
  </group>
</div>
`;
